### PR TITLE
Align validation of watchCacheSizes to `kube-apiserver` flag

### PR DESCRIPTION
**Pull request by tobschli**
**How to categorize this PR?**

/area robustness ipcei security
/kind enhancement

**What this PR does / why we need it**:

There are some special characters that we should disallow the user to specify, because this would result in a `kube-apiserver` in `CrashLoopBackOff`.

We should disallow the special characters that depict delimiters for this flag.

In addition to that, we should disallow capitalized strings, according to documentation. Even though the `kube-apiserver` starts when you specify a capitalized string.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```breaking user
It is not allowed anymore to specify the following characters: [,. #], as well as capitalized letters, within the entries of the Shoot's `.spec.kubernetes.kubeAPIServer.watchCacheSizes.resources[].resource` field. Please update your Shoots accordingly.
```

```breaking user
It is not allowed anymore to specify the following characters: [, #], as well as capitalized letters, within the entries of the Shoot's `.spec.kubernetes.kubeAPIServer.watchCacheSizes.resources[].apiGroup` field. Please update your Shoots accordingly.
```

```breaking operator
It is not allowed anymore to specify the following characters: [,. #], as well as capitalized letters, within the entries of the Garden's `.spec.virtualCluster.Gardener.apiServer.watchCacheSizes.resources[].resource` field. Please update your Gardens accordingly.
```

```breaking operator
It is not allowed anymore to specify the following characters: [, #], as well as capitalized letters, within the entries of the Garden's `.spec.virtualCluster.Gardener.apiServer.watchCacheSizes.resources[].apiGroup` field. Please update your Gardens accordingly.
```

### CODE REVIEW

**Review by marc1404 — COMMENTED** [review]
/lgtm

Thanks for matching our validation of `watchCacheSizes` with the expectations of the `kube-apiserver` flag!

I guess it might be a bit difficult to see that whitespace is a forbidden character, but this is acceptable.

File: pkg/apis/core/validation/shoot.go [hunk 1]
```diff
@@ -1103,16 +1103,39 @@ func validateNetworkingStatus(networking *core.NetworkingStatus, fldPath *field.
 // ValidateWatchCacheSizes validates the given WatchCacheSizes fields.
 func ValidateWatchCacheSizes(sizes *core.WatchCacheSizes, fldPath *field.Path) field.ErrorList {
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot.go:1104]: The doc string of the field says:
https://github.com/gardener/gardener/blob/3f8f06cae1899cc033f791794696d8d622c5e3fb/pkg/apis/core/types_shoot.go#L693-L694

I am wondering if it works at all setting the field to a non-zero value.

Comment (tobschli) [reply pkg/apis/core/validation/shoot.go:1104]: I am unsure what you mean by "works at all" 😅

When e.g. setting the default to `0` and a specific resource to something that is non-zero, our mechanisms correctly pass this into the `kube-apiserver` flags:

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: local
  namespace: garden-local
spec:
  ...
  kubernetes:
    kubeAPIServer:
      watchCacheSizes:
        default: 0
        resources:
          - resource: nodes
            size: 1
```

Resulting flags of `kube-apiserver`:
```
- --default-watch-cache-size=0
- --watch-cache-sizes=nodes#1
```

Comment (tobschli) [reply pkg/apis/core/validation/shoot.go:1104]: Wow, I just noticed, that `--default-watch-cache-size` is not a flag (anymore?).
When looking at the [cli-reference docs](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/), searching for this flag does not result in anything 😅

There is `--watch-cache` now, which is a `bool` that toggles the cache generally on or off.

I think we should adapt that in our code to switch to this flag instead, WDYT?

Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:1104]: You could check the source of k/k if the flag still exists. Occasionally, docs might be outdated. If the flag is removed, let's find out which PR removed it and when this happened. Maybe we have yet another case of obsolete field similar to https://github.com/gardener/gardener/issues/12830?

Comment (tobschli) [reply pkg/apis/core/validation/shoot.go:1104]: Yeah, I shortly checked k/k PRs without luck. Let me check source 😊

Comment (tobschli) [reply pkg/apis/core/validation/shoot.go:1104]: So, it actually seems to still be a flag ([ref](https://github.com/kubernetes/kubernetes/blob/21e2fcea9ec25c79d4ceea0c8a931c2655bf5d93/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L152-L153))
But it is a deprecated field.
Actually, when I think about it again, I would also expect the apiserver not to start due to unknown flags.

File: pkg/apis/core/validation/shoot.go [hunk 2]
```diff
@@ -1103,16 +1103,39 @@ func validateNetworkingStatus(networking *core.NetworkingStatus, fldPath *field.
...
 	allErrs := field.ErrorList{}
+
+	invalidCharacters := ",. #"
```
Comment (ScheererJ) [inline pkg/apis/core/validation/shoot.go:1107]: Do you think that an allow listing approach would be better suited instead of this block listing approach?

Comment (tobschli) [reply pkg/apis/core/validation/shoot.go:1107]: I thought about it, but had the concern, that an allow-list could easily be too restrictive, perhaps breaking many existing `Shoot`s.

Comment (ScheererJ) [reply pkg/apis/core/validation/shoot.go:1107]: Well, we basically allow all special characters except four. I would have expected that this to be mostly alphanumeric characters and some special characters like `-_/`.
However, having some validation is better than nothing. Hence, we can start with this for now.

**Comment by ialidzhikov** [comment]
I am sorry for the delay with the review on this PR. I was busy with a regular shift until Monday including. Let me have a look.

File: pkg/apis/core/validation/shoot.go (hunk 1 above)
Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:1104]: > You could check the source of k/k if the flag still exists. Occasionally, docs might be outdated. If the flag is removed, let's find out which PR removed it and when this happened. Maybe we have yet another case of obsolete field similar to https://github.com/gardener/gardener/issues/12830?

My understanding is:
- The deprecated `--default-watch-cache-size` flag is no-op. Ref https://github.com/kubernetes/kubernetes/blob/21e2fcea9ec25c79d4ceea0c8a931c2655bf5d93/staging/src/k8s.io/apiserver/pkg/server/options/etcd.go#L152-L153. The field to which the flag value is assigned is not used.
- The `--watch-cache-sizes` flag is not deprecated. Its help text is explanatory enough about which values make sense:
> [...]. The only meaningful size setting to supply here is zero, which means to disable watch caching for the associated resource; all non-zero values are equivalent and mean to not disable watch caching for that resource

tl;dr: We don't have to do something proactively. When Kubernetes removes the `--default-watch-cache-size` flag, then we have to remove the corresponding field.

File: pkg/apis/core/validation/shoot_test.go [hunk 3]
```diff
@@ -2379,6 +2379,22 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Required(field.NewPath("resources[0].resource"), "must not be empty"),
 					)),
+					Entry("invalid (core/resource name with casing", &core.WatchCacheSizes{
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
					Entry("invalid (resource with casing)", &core.WatchCacheSizes{
```

File: pkg/apis/core/validation/shoot_test.go [hunk 4]
```diff
@@ -2379,6 +2379,22 @@ var _ = Describe("Shoot Validation Tests", func() {
...
+						Resources: []core.ResourceWatchCacheSize{{
+							Resource:  "Secrets",
+							CacheSize: 42,
+						}},
+					}, ConsistOf(
+						field.Invalid(field.NewPath("resources[0].resource"), "Secrets", "must be lower case"),
+					)),
+					Entry("invalid (core/resource name with illegal character", &core.WatchCacheSizes{
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
					Entry("invalid (resource with illegal character)", &core.WatchCacheSizes{
```

File: pkg/apis/core/validation/shoot_test.go [hunk 5]
```diff
@@ -2412,6 +2428,24 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Required(field.NewPath("resources[0].resource"), "must not be empty"),
 					)),
+					Entry("invalid (Apps/deployments name with casing", &core.WatchCacheSizes{
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
					Entry("invalid (API group with casing)", &core.WatchCacheSizes{
```

File: pkg/apis/core/validation/shoot_test.go [hunk 6]
```diff
@@ -2412,6 +2428,24 @@ var _ = Describe("Shoot Validation Tests", func() {
...
+						Resources: []core.ResourceWatchCacheSize{{
+							APIGroup:  ptr.To("Apps"),
+							Resource:  "deployments",
+							CacheSize: 42,
+						}},
+					}, ConsistOf(
+						field.Invalid(field.NewPath("resources[0].apiGroup"), "Apps", "must be lower case"),
+					)),
+					Entry("invalid (core/resource name with illegal character", &core.WatchCacheSizes{
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
					Entry("invalid (API group with illegal character)", &core.WatchCacheSizes{
```

File: pkg/apis/core/validation/shoot_test.go [hunk 7]
```diff
@@ -2379,6 +2379,22 @@ var _ = Describe("Shoot Validation Tests", func() {
...
+						Resources: []core.ResourceWatchCacheSize{{
+							Resource:  "secrets#",
+							CacheSize: 42,
+						}},
+					}, ConsistOf(
+						field.Invalid(field.NewPath("resources[0].resource"), "secrets#", `must not contain any of the following characters: ",. #"`),
+					)),
 
 					// APIGroup set
 					Entry("valid (apps/deployments=0)", &core.WatchCacheSizes{
```
Comment (ialidzhikov) [inline pkg/apis/core/validation/shoot_test.go:2407]: Does it make sense to add a test case where API group is set to empty string? Just to cover the corner case that `apiGroup: ""` is valid (empty string is a valid value).

File: pkg/apis/core/validation/shoot_test.go (hunk 5 above)
Comment (tobschli) [reply pkg/apis/core/validation/shoot_test.go]: I tried to keep the naming form aligned to the other tests there ([ref](https://github.com/tobschli/gardener/blob/48d91cac9f0b39b84e22ebaa9f8faf2d66692951/pkg/apis/core/validation/shoot_test.go#L2400-L2428))
But yeah, the casing and the missing closing bracket is something I would fix 😅

File: pkg/apis/core/validation/shoot_test.go (hunk 7 above)
Comment (tobschli) [reply pkg/apis/core/validation/shoot_test.go:2407]: Does not hurt, I guess. Will do :)

File: pkg/apis/core/validation/shoot_test.go (hunk 5 above)
Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot_test.go]: I see. But I don't think this naming pattern is applicable for all the test cases we have and we introduce with this PR.

If we take a look at the below snippet from the current revision of the PR:

```golang
					Entry("invalid (core/resource name with illegal character)", &core.WatchCacheSizes{
						Resources: []core.ResourceWatchCacheSize{{
							APIGroup:  ptr.To("apps#"),
							Resource:  "deployments",
							CacheSize: 42,
						}},
					}, ConsistOf(
						field.Invalid(field.NewPath("resources[0].apiGroup"), "apps#", `must not contain any of the following characters: ",. #"`),
					)),
```

it is not about the core group and it is the API group that has special character, not the resource.

If we take a look at:
```golang
					Entry("invalid (apps/deployments name with casing)", &core.WatchCacheSizes{
						Resources: []core.ResourceWatchCacheSize{{
							APIGroup:  ptr.To("Apps"),
							Resource:  "deployments",
							CacheSize: 42,
						}},
					}, ConsistOf(
						field.Invalid(field.NewPath("resources[0].apiGroup"), "Apps", "must be lower case"),
					)),
```

it is not clear from the entry description which has a name with a casing - the API group or the resource?

---


My full suggestion:
```diff
diff --git a/pkg/apis/core/validation/shoot_test.go b/pkg/apis/core/validation/shoot_test.go
index dc714a5c4d..c03762232b 100644
--- a/pkg/apis/core/validation/shoot_test.go
+++ b/pkg/apis/core/validation/shoot_test.go
@@ -2357,7 +2357,7 @@ var _ = Describe("Shoot Validation Tests", func() {
 							CacheSize: 0,
 						}},
 					}, BeEmpty()),
-					Entry("valid (\"\"/secrets=0)", &core.WatchCacheSizes{
+					Entry("valid (API group empty)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
 							APIGroup:  ptr.To(""),
 							Resource:  "secrets",
@@ -2378,7 +2378,7 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Invalid(field.NewPath("resources[0].size"), int64(negativeSize), apivalidation.IsNegativeErrorMsg).WithOrigin("minimum"),
 					)),
-					Entry("invalid (core/resource empty)", &core.WatchCacheSizes{
+					Entry("invalid (core resource empty)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
 							Resource:  "",
 							CacheSize: 42,
@@ -2386,7 +2386,7 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Required(field.NewPath("resources[0].resource"), "must not be empty"),
 					)),
-					Entry("invalid (core/resource name with casing)", &core.WatchCacheSizes{
+					Entry("invalid (core resource with casing)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
 							Resource:  "Secrets",
 							CacheSize: 42,
@@ -2394,7 +2394,7 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Invalid(field.NewPath("resources[0].resource"), "Secrets", "must be lower case"),
 					)),
-					Entry("invalid (core/resource name with illegal character)", &core.WatchCacheSizes{
+					Entry("invalid (core resource with illegal character)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
 							Resource:  "secrets#",
 							CacheSize: 42,
@@ -2427,15 +2427,16 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Invalid(field.NewPath("resources[0].size"), int64(negativeSize), apivalidation.IsNegativeErrorMsg).WithOrigin("minimum"),
 					)),
-					Entry("invalid (apps/resource empty)", &core.WatchCacheSizes{
+					Entry("invalid (apps resource empty)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
+							APIGroup:  ptr.To("apps"),
 							Resource:  "",
 							CacheSize: 42,
 						}},
 					}, ConsistOf(
 						field.Required(field.NewPath("resources[0].resource"), "must not be empty"),
 					)),
-					Entry("invalid (apps/deployments name with casing)", &core.WatchCacheSizes{
+					Entry("invalid (apps API group with casing)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
 							APIGroup:  ptr.To("Apps"),
 							Resource:  "deployments",
@@ -2444,7 +2445,7 @@ var _ = Describe("Shoot Validation Tests", func() {
 					}, ConsistOf(
 						field.Invalid(field.NewPath("resources[0].apiGroup"), "Apps", "must be lower case"),
 					)),
-					Entry("invalid (core/resource name with illegal character)", &core.WatchCacheSizes{
+					Entry("invalid (apps API group with illegal character)", &core.WatchCacheSizes{
 						Resources: []core.ResourceWatchCacheSize{{
 							APIGroup:  ptr.To("apps#"),
 							Resource:  "deployments",

```

File: pkg/apis/core/validation/shoot.go (hunk 1 above)
Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot.go:1104]: For the record, I created https://github.com/kubernetes/kubernetes/pull/134151 to make it obvious that the `--default-watch-cache-size` is no-op.

File: pkg/apis/core/validation/shoot_test.go (hunk 5 above)
Comment (tobschli) [reply pkg/apis/core/validation/shoot_test.go]: In the end, I don't have such a strong opinion about test naming.
Will update

Comment (ialidzhikov) [reply pkg/apis/core/validation/shoot_test.go]: With https://github.com/gardener/gardener/pull/12839/commits/34cdea5b0dc91d8067d3ed70300b78529380ebf7, not all suggestions from https://github.com/gardener/gardener/pull/12839#discussion_r2361997081 were applied. I pushed f72db1a6631f890dace8159b78ac2251136ae9a3 to apply all of them.
Let me know if you have comments/objections.

**Review by ialidzhikov — APPROVED** [review]
/lgtm
